### PR TITLE
E2E tests for manifest signing set up

### PR DESCRIPTION
**Pull request by ggjjj**
E2E set up for Manifest Signing Feature
1. Tests 3 different scenarios by configuring Edge Daemon with a good root ca, bad root ca, and no root ca
2. Building manifest signer client in the pipeline and using it in the test framework to use the signed manifest 
3. Contains ManifestSignerTest fixture which sets the E2E

### CODE REVIEW

**Comment by vadim-kovalyov** [comment]
Closing old draft pr.

File: builds/e2e/e2e.yaml [hunk 1]
```diff
@@ -32,28 +32,28 @@ variables:
 jobs:
 
 ################################################################################
-  - job: linux_arm32v7
+# - job: linux_arm32v7
```
Comment (darobs) [inline builds/e2e/e2e.yaml:35]: Please be sure to uncomment this and 82-106

File: builds/e2e/e2e.yaml [hunk 2]
```diff
@@ -156,65 +157,65 @@ jobs:
     - template: templates/e2e-run.yaml
 
 ################################################################################
-  - job: centos7_amd64
+  #- job: centos7_amd64
```
Comment (darobs) [inline builds/e2e/e2e.yaml:160]: PLease be sure to uncomment this and 190-221

File: builds/e2e/templates/e2e-run.yaml [hunk 3]
```diff
@@ -18,65 +18,68 @@ steps:
...
-
-      # Below tests were disabled and marked for re-enable when a blocking item was resolved.
-      # When it was resolved the tests were never enabled. We need to re-enable these.
-      $filter += '&FullyQualifiedName!~Provisioning&FullyQualifiedName!~SasOutOfScope&FullyQualifiedName!~X509ManualProvision&FullyQualifiedName!~AuthorizationPolicyUpdateTest&FullyQualifiedName!~AuthorizationPolicyExplicitPolicyTest'
-    }
-    elseif ($test_type -eq 'nestededge_amqp')
-    {
-      $filter += '&Category!=SingleNodeOnly'
-      $filter += '&Category!=BrokerRequired'
+    # Filter out Manifest Signing tests.
```
Comment (darobs) [inline builds/e2e/templates/e2e-run.yaml:21]: Please make sure to undo these changes, as well.

File: builds/e2e/templates/e2e-setup.yaml [hunk 4]
```diff
@@ -77,6 +82,35 @@ steps:
...
+    $manifestSignerClientBinDir = Convert-Path "$manifestSignerClientDir/bin/Debug/netcoreapp3.1"
+    Write-Output "##vso[task.setvariable variable=manifestSignerClientDir]$manifestSignerClientDir"
+    Write-Output "##vso[task.setvariable variable=manifestSignerClientBinDir]$manifestSignerClientBinDir"
+    $manifestSigningTestDir = '$(System.ArtifactsDirectory)/manifest_signing'
+    $manifestSigningDeploymentDir = '$(System.ArtifactsDirectory)/manifest_signing/deployment'
+    Write-Output "##vso[task.setvariable variable=manifestSigningDeploymentDir]$manifestSigningDeploymentDir"
+    $manifestSigningCertDir = '$(System.ArtifactsDirectory)/manifest_signing/certs'
+    New-Item "$manifestSigningCertDir" -ItemType Directory -Force | Out-Null
+    New-Item "$manifestSigningDeploymentDir" -ItemType Directory -Force | Out-Null
+    New-Item -Path "$manifestSigningDeploymentDir/deployment.json" -ItemType File
```
Comment (darobs) [inline builds/e2e/templates/e2e-setup.yaml:97]: Do we need to create the deployment.json and signed_deployment.json file here?  This doesn't seem to be hurting anything, but is it necessary?

File: builds/e2e/templates/e2e-setup.yaml [hunk 5]
```diff
@@ -117,6 +151,46 @@ steps:
...
+          }
+        }
+      };
+
+      $testManifestSigningDefaultLaunchSettings = $testManifestSigningDefaultLaunchSettings | ConvertTo-Json -Depth 3 
+      ##$testManifestSigningDefaultLaunchSettings | ConvertTo-Json -Depth 3 | Out-File -Encoding Utf8 "$testManifestSigningLaunchSettingsPath"
+      $testManifestSigningDefaultLaunchSettings | Out-File -Encoding Utf8 "$testManifestSigningLaunchSettingsPath"
+      echo "LaunchSettingPath is"
+      Get-Content -Path "$testManifestSigningLaunchSettingsPath"
+      Get-Content -Path "$testManifestSigningSignerCertPath"
```
Comment (darobs) [inline builds/e2e/templates/e2e-setup.yaml:189]: Some cleanup is required here, but you might want to see the launch settings if we have to debug this test later.

File: builds/e2e/templates/e2e-setup.yaml [hunk 6]
```diff
@@ -184,4 +271,5 @@ steps:
     }
 
     $context | ConvertTo-Json | Out-File -Encoding Utf8 '$(binDir)/context.json'
+    Get-Content -Path '$(binDir)/context.json'
```
Comment (darobs) [inline builds/e2e/templates/e2e-setup.yaml:274]: You should probably remove this.

File: samples/dotnet/ManifestSignerClient/Program.cs [hunk 7]
```diff
@@ -242,7 +242,7 @@ static void Main()
             catch (Exception e)
             {
                 Console.WriteLine(e.ToString());
-                Environment.Exit(0);
+                // Environment.Exit(0);
```
Comment (darobs) [inline samples/dotnet/ManifestSignerClient/Program.cs:245]: Should this exit w/ a non-zero value on failure?

File: test/Microsoft.Azure.Devices.Edge.Test.Common/EdgeRuntime.cs [hunk 8]
```diff
@@ -61,9 +67,37 @@ public EdgeRuntime(string deviceId, Option<string> agentImage, Option<string> hu
...
-
             DateTime deployTime = DateTime.Now;
             EdgeConfiguration edgeConfiguration = builder.Build();
+            string signedConfig = string.Empty;
+
+            ProcessStartInfo startInfo;
+
+            if (enableManifestSigning.HasValue)
+            {
+                // Wrtie the current config into a file: EdgeConfiguration ToString() outputs the ConfigurationContent
```
Comment (darobs) [inline test/Microsoft.Azure.Devices.Edge.Test.Common/EdgeRuntime.cs:78]: nit: typo

File: test/Microsoft.Azure.Devices.Edge.Test.Common/EdgeRuntime.cs [hunk 9]
```diff
@@ -61,9 +67,37 @@ public EdgeRuntime(string deviceId, Option<string> agentImage, Option<string> hu
...
+                signedConfig = File.ReadAllText(signedDeploymentPath);
+            }
+
+            if (!string.IsNullOrEmpty(signedConfig))
+            {
+                // Convert signed config to ConfigurationContent
+                edgeConfiguration.Config = JsonConvert.DeserializeObject<ConfigurationContent>(signedConfig);
+            }
+
+            Console.WriteLine($"\n Signed config str = {signedConfig}");
```
Comment (darobs) [inline test/Microsoft.Azure.Devices.Edge.Test.Common/EdgeRuntime.cs:100]: Probably not needed

File: test/Microsoft.Azure.Devices.Edge.Test.Common/certs/CaCertificates.cs [hunk 10]
```diff
@@ -44,5 +45,17 @@ public CaCertificates(string certificatePath, string keyPath, string trustedCert
             this.KeyPath = keyPath;
             this.TrustedCertificatesPath = trustedCertsPath;
         }
+
+        public CaCertificates(string certificatePath, string keyPath, string trustedCertsPath, string manifestSigningTrustBundlePath)
+        {
+            Preconditions.CheckArgument(File.Exists(certificatePath));
+            Preconditions.CheckArgument(File.Exists(keyPath));
+            Preconditions.CheckArgument(File.Exists(trustedCertsPath));
+            // Preconditions.CheckArgument(File.Exists(manifestSigningTrustBundlePath));
```
Comment (darobs) [inline test/Microsoft.Azure.Devices.Edge.Test.Common/certs/CaCertificates.cs:54]: remove commented code if not needed

File: test/Microsoft.Azure.Devices.Edge.Test/ManifestSigning.cs [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+        public async Task TestIfSignedDeploymentIsConfiguredWithBadRootCa()
+        {
+            // Edge Daemon is configured with a bad root CA but manifest is signed.
+            this.SetLaunchSettingsWithRootCa(Context.Current.ManifestSigningDefaultLaunchSettings, Context.Current.ManifestSigningGoodRootCaPath);
+            ManifestSettings inputManifestSettings = new ManifestSettings(Context.Current.ManifestSigningDeploymentPath, Context.Current.ManifestSigningSignedDeploymentPath, Context.Current.ManifestSigningGoodRootCaPath, Context.Current.ManifestSignerClientDirectory, Context.Current.ManifestSignerClientProjectPath);
+
+            await this.SetConfigToEdgeDaemon(Context.Current.ManifestSigningBadRootCaPath, this.TestToken);
+
+            // Set a faster time out as its expected to fail.
+            CancellationTokenSource manifestSigningCts = new CancellationTokenSource(TimeSpan.FromMinutes(5));
```
Comment (darobs) [inline test/Microsoft.Azure.Devices.Edge.Test/ManifestSigning.cs:156]: I'd set this to 2 minutes (or less, if you can get away with it).

File: test/Microsoft.Azure.Devices.Edge.Test/ManifestSigning.cs [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+                        },
+                        manifestSigningCts.Token,
+                        Context.Current.NestedEdge,
+                        inputManifestSettings);
+                }
+                catch (TaskCanceledException)
+                {
+                    CancellationTokenSource getTwinTimer = new CancellationTokenSource(TimeSpan.FromMinutes(5));
+                    Twin twin = await this.IotHub.GetTwinAsync(this.runtime.DeviceId, Option.Some("$edgeAgent"), getTwinTimer.Token);
+                    Assert.AreNotEqual(twin.Properties.Desired.Version, twin.Properties.Reported.GetLastUpdatedVersion());
```
Comment (darobs) [inline test/Microsoft.Azure.Devices.Edge.Test/ManifestSigning.cs:217]: Like in the test above, we should move this out of the test block.

Comment (darobs) [reply test/Microsoft.Azure.Devices.Edge.Test/ManifestSigning.cs:217]: Also, I wonder if we should put a comment here:
// EdgeAgent will reject the new deployment, so the reported and desired versions will not match,

File: test/Microsoft.Azure.Devices.Edge.Test/SetupFixture.cs [hunk 13]
```diff
@@ -227,5 +228,31 @@ public void AddCertsToConfig(DaemonConfiguration config)
...
+            File.Copy(this.certs.TrustedCertificatesPath, trustBundlePath);
+            OsPlatform.Current.SetOwner(trustBundlePath, "aziotcs", "644");
+            File.Copy(this.certs.CertificatePath, certPath);
+            OsPlatform.Current.SetOwner(certPath, "aziotcs", "644");
+            File.Copy(this.certs.KeyPath, keyPath);
+            OsPlatform.Current.SetOwner(keyPath, "aziotks", "600");
+
+            if (inputManifestSigningTrustBundlePath.HasValue)
+            {
+                // File.Create(manifestSigningTrustBundlePath);
```
Comment (darobs) [inline test/Microsoft.Azure.Devices.Edge.Test/SetupFixture.cs:250]: Remove commented code.

File: test/Microsoft.Azure.Devices.Edge.Test/helpers/BaseFixture.cs [hunk 14]
```diff
@@ -61,6 +60,9 @@ protected async Task AfterEachTestAsync()
                     }
                 },
                 "Completed test teardown");
+
+            await this.AfterTestTimerEnds();
+            this.cts.Dispose();
```
Comment (darobs) [inline test/Microsoft.Azure.Devices.Edge.Test/helpers/BaseFixture.cs:65]: Check with E2E test experts that its OK to dispose of `this.cts` later.

File: test/Microsoft.Azure.Devices.Edge.Test/helpers/ManifestSigningSetupFixture.cs [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+        protected override Task AfterTestTimerEnds() => this.daemon.ConfigureAsync(
+                        config =>
+                        {
+                            config.RemoveManifestTrustBundle();
+
+                            config.Update();
+                            return Task.FromResult((
+                                "Removed Manifest Trust Bundle", new object[] { }));
+                        },
+                        this.TestToken);
```
Comment (darobs) [inline test/Microsoft.Azure.Devices.Edge.Test/helpers/ManifestSigningSetupFixture.cs:23]: Maybe this should use its own token.

**Review by darobs — COMMENTED** [review]
Overall this is shaping up well, needs some cleanup before it can be submitted.

**Review by darobs — DISMISSED** [review]
Barring comments  from @yophilav (or other E2E experts), ship it!

**Review by darobs — APPROVED** [review]
👍

File: test/Microsoft.Azure.Devices.Edge.Test.Common/DaemonConfiguration.cs [hunk 16]
```diff
@@ -265,6 +270,8 @@ public void RemoveCertificates()
             this.config[Service.Certd].Document.RemoveIfExists("preloaded_certs.aziot-edged-trust-bundle");
         }
 
+        public void RemoveManifestTrustBundle() => this.config[Service.Certd].Document.RemoveIfExists("preloaded_certs.aziot-edged-manifest-trust-bundle");
```
[the remark this answers is not in the document]
Comment (ggjjj) [reply test/Microsoft.Azure.Devices.Edge.Test.Common/DaemonConfiguration.cs:273]: yes

File: builds/e2e/e2e.yaml [hunk 17]
```diff
@@ -46,6 +46,7 @@ jobs:
       artifactName: iotedged-debian9-arm32v7
       identityServiceArtifactName: packages_debian-9-slim_arm32v7
       identityServicePackageFilter: aziot-identity-service_*_armhf.deb
+      enableManifestSigning: false
```
[the remark this answers is not in the document]
Comment (ggjjj) [reply builds/e2e/e2e.yaml:49]: Actually it should be set to true. My bad. fixed it.

File: builds/e2e/templates/e2e-setup.yaml [hunk 18]
```diff
@@ -20,7 +20,12 @@ steps:
       TestRootCaCertificate,
       TestRootCaKey,
       TestRootCaPassword,
-      TestBlobStoreSas
+      TestBlobStoreSas,
+      TestManifestSigningGoodRootCa,
+      TestManifestSigningIntermediateCa,
+      TestManifestSigningSignerCert,
+      TestManifestSigningSignerKey,
+      TestManifestSigningBadRootCa
```
[the remark this answers is not in the document]
Comment (ggjjj) [reply builds/e2e/templates/e2e-setup.yaml:28]: I have set the keyvault secret expiration dates. Also the documentation to generate these certs are in the https://github.com/Azure/iotedge/blob/master/samples/dotnet/ManifestSignerClient/Readme.md

File: test/Microsoft.Azure.Devices.Edge.Test.Common/EdgeRuntime.cs [hunk 19]
```diff
@@ -61,9 +67,36 @@ public EdgeRuntime(string deviceId, Option<string> agentImage, Option<string> hu
...
+            ProcessStartInfo startInfo;
+
+            if (enableManifestSigning.HasValue)
+            {
+                // Write the current config into a file: EdgeConfiguration ToString() outputs the ConfigurationContent
+                string deploymentPath = enableManifestSigning.OrDefault().ManifestSigningDeploymentPath.OrDefault();
+                File.WriteAllText(deploymentPath, edgeConfiguration.ToString());
+
+                // Run Manifest signer client
+                string projectDirectory = enableManifestSigning.OrDefault().ManifestSignerClientProjectPath.OrDefault();
```
[the remark this answers is not in the document]
Comment (ggjjj) [reply test/Microsoft.Azure.Devices.Edge.Test.Common/EdgeRuntime.cs:81]: I am already doing the same but it just the manifest signer client gets executed with `dotnet run` command and picks up the launchsettings.json. I will look for doing it better and make changes in later PR

File: test/Microsoft.Azure.Devices.Edge.Test/ManifestSigning.cs [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+                File.WriteAllText(Context.Current.ManifestSigningLaunchSettingsPath.OrDefault(), defaultJsonObject.ToString());
+                string newLauchSettingsContents = File.ReadAllText(Context.Current.ManifestSigningLaunchSettingsPath.OrDefault());
+                Console.WriteLine($"new Launch Settings from SetLaunchSettingsWithRootCa: {newLauchSettingsContents} ");
+            }
+        }
+
+        [Test]
+        public async Task TestIfSignedDeploymentIsSuccessful()
+        {
+            if (Context.Current.EnableManifestSigning)
```
[the remark this answers is not in the document]
Comment (ggjjj) [reply test/Microsoft.Azure.Devices.Edge.Test/ManifestSigning.cs:76]: I will remove the flag altogether as keep the barrier of signing in `DeployConfigurationAsync`

File: test/Microsoft.Azure.Devices.Edge.Test.Common/EdgeRuntime.cs [hunk 21]
```diff
@@ -61,9 +67,36 @@ public EdgeRuntime(string deviceId, Option<string> agentImage, Option<string> hu
...
+
+                // write the signed deployment into a file
+                string signedDeploymentPath = enableManifestSigning.OrDefault().ManifestSigningSignedDeploymentPath.OrDefault();
+                signedConfig = File.ReadAllText(signedDeploymentPath);
+            }
+
+            if (!string.IsNullOrEmpty(signedConfig))
+            {
+                // Convert signed config to ConfigurationContent
+                edgeConfiguration.Config = JsonConvert.DeserializeObject<ConfigurationContent>(signedConfig);
```
[the remark this answers is not in the document]
Comment (ggjjj) [reply test/Microsoft.Azure.Devices.Edge.Test.Common/EdgeRuntime.cs:103]: as discussed, I will handle them better in another PR. For now, this is the best way I have found to handle singed deployment content

**Review by damonbarry — APPROVED** [review]
